File: src/shapes/rapier_segment_shape_2d.cpp
#include "rapier_segment_shape_2d.h"

rapier2d::Handle RapierSegmentShape2D::create_rapier_shape() const {
	Vector2 direction = b - a;
	direction.normalize();

	Vector2 perpendicular = Vector2(-direction.y, direction.x);
	double height = 0.1;

	Vector2 p1 = a + perpendicular * height / 2.0;
	Vector2 p2 = a - perpendicular * height / 2.0;
	Vector2 p3 = b + perpendicular * height / 2.0;
	Vector2 p4 = b - perpendicular * height / 2.0;

	rapier2d::Vector rapier_points[4];
	rapier_points[0] = rapier2d::Vector{ p1.x, p1.y };
	rapier_points[1] = rapier2d::Vector{ p2.x, p2.y };
	rapier_points[2] = rapier2d::Vector{ p3.x, p3.y };
	rapier_points[3] = rapier2d::Vector{ p4.x, p4.y };

	return rapier2d::shape_create_convex_polyline(rapier_points, 4);
}

void RapierSegmentShape2D::set_data(const Variant &p_data) {
	ERR_FAIL_COND(p_data.get_type() != Variant::RECT2);

	Rect2 r = p_data;
	a = r.position;
	b = r.size;
	n = (b - a).orthogonal();

	Rect2 aabb;
	aabb.position = a;
	aabb.expand_to(b);
	if (aabb.size.x == 0) {
		aabb.size.x = 0.001;
	}
	if (aabb.size.y == 0) {
		aabb.size.y = 0.001;
	}
	configure(aabb);
}

Variant RapierSegmentShape2D::get_data() const {
	Rect2 r;
	r.position = a;
	r.size = b;
	return r;
}

real_t RapierSegmentShape2D::get_moment_of_inertia(real_t p_mass, const Size2 &p_scale) const {
	return p_mass * ((a * p_scale).distance_squared_to(b * p_scale)) / 12.0;
}

File: src/shapes/rapier_segment_shape_2d.rs
use godot::classes::physics_server_2d::ShapeType;
use godot::prelude::*;

use super::rapier_shape::RapierShape;
use crate::rapier_wrapper::prelude::*;
use crate::servers::rapier_physics_singleton::insert_id_rid;
use crate::servers::rapier_physics_singleton::PhysicsIds;
use crate::servers::rapier_physics_singleton::PhysicsShapes;
use crate::shapes::rapier_shape::IRapierShape;
use crate::shapes::rapier_shape_base::RapierShapeBase;
pub struct RapierSegmentShape2D {
    base: RapierShapeBase,
}
impl RapierSegmentShape2D {
    pub fn create(rid: Rid, physics_shapes: &mut PhysicsShapes, physics_ids: &mut PhysicsIds) {
        let shape = Self {
            base: RapierShapeBase::new(rid),
        };
        insert_id_rid(shape.base.get_id(), rid, physics_ids);
        physics_shapes.insert(rid, RapierShape::RapierSegmentShape2D(shape));
    }
}
impl IRapierShape for RapierSegmentShape2D {
    fn get_base(&self) -> &RapierShapeBase {
        &self.base
    }

    fn get_mut_base(&mut self) -> &mut RapierShapeBase {
        &mut self.base
    }

    fn get_type(&self) -> ShapeType {
        ShapeType::SEGMENT
    }

    fn allows_one_way_collision(&self) -> bool {
        true
    }

    fn set_data(&mut self, data: Variant, physics_engine: &mut PhysicsEngine) {
        if data.get_type() != VariantType::RECT2 {
            godot_error!("RapierSegmentShape data must be a Rect2. Got {}", data);
            return;
        }
        let r: Rect2 = data.try_to().unwrap_or_default();
        let p1 = r.position;
        let p2 = r.size;
        let rapier_points = [vector_to_rapier(p1), vector_to_rapier(p2)];
        let handle = physics_engine.shape_create_concave_polyline(&rapier_points.to_vec(), None);
        self.base.set_handle_and_reset_aabb(handle, physics_engine);
    }

    fn get_data(&self, physics_engine: &PhysicsEngine) -> Variant {
        let (points, _) = physics_engine.shape_get_concave_polyline(self.base.get_handle());
        if points.len() != 2 {
            godot_error!(
                "RapierSegmentShape data must be a Rect2. Got length {}",
                points.len()
            );
            return Rect2::default().to_variant();
        }
        let r = Rect2::new(
            vector_to_godot(points[0].coords),
            vector_to_godot(points[1].coords),
        );
        r.to_variant()
    }
}
